src/lucky_cli/task.cr

```diff
@@ -60,6 +68,131 @@ abstract class LuckyCli::Task
     end
   end
 
+  macro positional_arg(arg_name, description, required = nil, format = nil, to = nil)
+    {% PARSER_OPTS << arg_name %}
+    @{{ arg_name.id }} : {% if to %}Array(String){% else %}String{% end %} | Nil
+
+    def set_opt_for_{{ arg_name.id }}(args : Array(String))
+      {% if to %}
+        value = args[@positional_arg_count..-1]
+      {% else %}
+        value = args[@positional_arg_count]?
+      {% end %}
+      {% if format %}
+      matches = value.is_a?(Array) ? value.all?(&.=~({{ format }})) : value =~ {{ format }}
+      if matches
+        @{{ arg_name.id }} = value
+      else
+        raise "Invalid format for {{ arg_name.id }}. It should match {{ format }}"
+      end
+      {% else %}
+        @{{ arg_name.id }} = value
+      {% end %}
+      @positional_arg_count += 1
+    end
+
+    def {{ arg_name.id }}
+      {% if required %}
+        if @{{ arg_name.id }}.nil?
+          raise <<-ERROR
+          {{ arg_name.id }} is required, but no value was passed.
+          ERROR
+        end
+        @{{ arg_name.id }}.not_nil!
+      {% else %}
+        @{{ arg_name.id }}
+      {% end %}
+    end
+  end
+
+  macro arg(arg_name, description, shortcut = nil, required = nil, format = nil)
+    {% PARSER_OPTS << arg_name %}
+    @{{ arg_name.id }} : String?
+
+    def set_opt_for_{{ arg_name.id }}(unused_args : Array(String))
+      {% if shortcut %}
+      option_parser.on(
+        "{{ shortcut.id }} {{ arg_name.stringify.upcase.id }}",
+        "--{{ arg_name.id.stringify.underscore.gsub(/_/, "-").id }}={{ arg_name.id.stringify.upcase.id }}",
+        {{ description }}
+      ) do |value|
+        value = value.strip
+        {% if format %}
+        if value =~ {{ format }}
+          @{{ arg_name.id }} = value
+        else
+          raise "Invalid format for {{ arg_name.id }}. It should match {{ format }}"
+        end
+        {% else %}
+          @{{ arg_name.id }} = value
+        {% end %}
+      end
+      {% else %}
+      option_parser.on(
+        "--{{ arg_name.id.stringify.underscore.gsub(/_/, "-").id }}={{ arg_name.id.stringify.upcase.id }}",
+        {{ description }}
+      ) do |value|
+        value = value.strip
+        {% if format %}
+        if value =~ {{ format }}
+          @{{ arg_name.id }} = value
+        else
+          raise "Invalid format for {{ arg_name.id }}. It should match {{ format }}"
+        end
+        {% else %}
+          @{{ arg_name.id }} = value
+        {% end %}
+      end
+      {% end %}
+    end
+
+    def {{ arg_name.id }} : String{% if !required %}?{% end %}
+      {% if required %}
```

Comment:
Same comment as above... wondering if the required validation and exception should run when setting the value instead of reading it. I didn't do a code example because it would muddy the previous comment about drying up `set_opt_for_...`.

Reply:
I wanted to do this, but ran in to an issue. Maybe you can help me through it though? For `arg`, you don't know if the value is set until after the options have been parsed. The parsed method isn't called until after they've already been set. For example:

```crystal
@val = nil
parser.on("--x=X", "") do |val|
  @val = val
  #if you get here, `val` has a value
end
@val == nil
parser.parse(ARGV)
# now you know if @val is nil or not, but the validation check would have to be here
```

This would mean I'd have to add a new method to run post parsing validations (which I'm doing now, but just in the form of you calling those methods). Maybe that's ok? 🤷‍♂

Reply:
Yeah, I think it's ok. I had looked at the positional_arg macro and it works their, but I hadn't considered in the arg macro it wasn't possible. I jumped through a few mental contortions to try to figure it out, but it seems to work well as it is. 😄

Reply:
I agree it'd be nice and I'm also ok shipping as -is. Maybe we come up with something later!